mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -6255,10 +6295,70 @@ struct ThreadwiseCopyRewritePattern
         else
           destLowerIndices = destUpperIndices;
 
+        bool toEmitOOBStoreCheckLogic = false;
+        SmallVector<unsigned, 2> oobStoreCheckDims;
+        if (destLowerIndices.size() == 5) {
+          for (unsigned iter = 0; iter < destTransform.getNumResults();
+               ++iter) {
+            auto expr = destTransform.getResult(iter);
+            if (iter == 2) { // put gemmM or gemmN dim
+              toEmitOOBStoreCheckLogic = true;
+              oobStoreCheckDims.push_back(iter);
+            }
+          }
+        }
+
         // Store to dest.
         // Issue scalar store.
-        b.create<StoreOp>(loc, convertedScalarValue, op.dest(),
-                          destLowerIndices);
+        if (toEmitOOBStoreCheckLogic) {
+          Value withinStoreBoundsOp =
+              b.create<ConstantIntOp>(loc, 1, b.getIntegerType(1));
+
+          auto iv0 = b.create<IndexCastOp>(loc, destLowerIndices[0],
+                                           b.getIntegerType(32));
+          auto iv1 = b.create<IndexCastOp>(loc, destLowerIndices[1],
+                                           b.getIntegerType(32));
+          auto iv2 = b.create<IndexCastOp>(loc, destLowerIndices[2],
+                                           b.getIntegerType(32));
+
+          Value oobAddrOp =
+              b.create<ConstantIntOp>(loc, 2147483648, b.getIntegerType(32));
+
+          auto iv3 = b.create<IndexCastOp>(loc, destLowerIndices[3],
+                                           b.getIntegerType(32));
+          auto iv4 = b.create<IndexCastOp>(loc, destLowerIndices[4],
+                                           b.getIntegerType(32));
+          // 2G ,INT MAX Value = 2147483647, use 2147483648 as offset and buffer
+          // store do nothing
+
+          auto coordLast = destLowerIndices[2];
+          auto sevenConstantOp = b.create<ConstantIndexOp>(loc, 3);
+
+          // coordLast<7, the last PR will change to real oob check
+          Value oobCheckOp = b.create<CmpIOp>(loc, CmpIPredicate::slt,
+                                              coordLast, sevenConstantOp);
+          auto ifWithinBoundsOp = b.create<scf::IfOp>(
+              loc,
+              TypeRange{b.getIntegerType(32), b.getIntegerType(32),
+                        b.getIntegerType(32), b.getIntegerType(32),
+                        b.getIntegerType(32)},
+              // withinStoreBoundsOp, true);
+              oobCheckOp, true);
+
+          auto thenBuilder = ifWithinBoundsOp.getThenBodyBuilder();
+          thenBuilder.create<scf::YieldOp>(loc,
+                                           ValueRange{iv0, iv1, iv2, iv3, iv4});
+          auto elseBuilder = ifWithinBoundsOp.getElseBodyBuilder();
+          elseBuilder.create<scf::YieldOp>(
+              loc, ValueRange{iv0, iv1, oobAddrOp, iv3, iv4});
+          // loc, ValueRange{iv0,iv1,oobAddrOp,iv3,iv4});
+
+          b.create<gpu::MubufStoreOp>(loc, convertedScalarValue, op.dest(),
```

Comment:
@asleepzzz  per discussion, we are suspecting the symptoms you saw is potentially due to wrapping around when we linearize the offset.

to continue look into the issue, please do:

1) change host code so it populates predefined non-zero constants onto the output tensor.
2) make sure tuning parameters are set consistently.
3) change GPU dialect syntax of `MubufStoreOp` so we supply another operand to decide whether to add a 2G offset.
4) change GPU->ROCDL lowering pipeline to use the operand and add a 2G offset in case needed.